src/voice/gateway/gateway.rs

```diff
@@ -0,0 +1,339 @@
+        trace!("Created voice socket url: {}", processed_url.clone());
+
+        let (websocket_send, mut websocket_receive) =
+            WebSocketBackend::connect(&processed_url).await?;
+
+        let shared_websocket_send = Arc::new(Mutex::new(websocket_send));
+
+        // Create a shared broadcast channel for killing all gateway tasks
+        let (kill_send, mut _kill_receive) = tokio::sync::broadcast::channel::<()>(16);
+
+        // Wait for the first hello and then spawn both tasks so we avoid nested tasks
+        // This automatically spawns the heartbeat task, but from the main thread
+        #[cfg(not(target_arch = "wasm32"))]
+        let msg: VoiceGatewayMessage = websocket_receive.next().await.unwrap().unwrap().into();
+        #[cfg(target_arch = "wasm32")]
+        let msg: VoiceGatewayMessage = websocket_receive.next().await.unwrap().into();
+        let gateway_payload: VoiceGatewayReceivePayload = serde_json::from_str(&msg.0).unwrap();
+
+        if gateway_payload.op_code != VOICE_HELLO {
+            return Err(VoiceGatewayError::NonHelloOnInitiate {
+                opcode: gateway_payload.op_code,
+            });
+        }
+
+        info!("VGW: Received Hello");
+
+        // The hello data for voice gateways is in float milliseconds, so we convert it to f64 seconds
+        let gateway_hello: VoiceHelloData =
+            serde_json::from_str(gateway_payload.data.get()).unwrap();
+        let heartbeat_interval_seconds: f64 = gateway_hello.heartbeat_interval / 1000.0;
+
+        let voice_events = VoiceEvents::default();
+        let shared_events = Arc::new(Mutex::new(voice_events));
+
+        let mut gateway = VoiceGateway {
+            events: shared_events.clone(),
+            heartbeat_handler: VoiceHeartbeatHandler::new(
+                Duration::from_secs_f64(heartbeat_interval_seconds),
+                1, // to:do actually compute nonce
+                shared_websocket_send.clone(),
+                kill_send.subscribe(),
+            ),
+            websocket_send: shared_websocket_send.clone(),
+            websocket_receive,
+            kill_send: kill_send.clone(),
+        };
+
+        // Now we can continuously check for messages in a different task, since we aren't going to receive another hello
+        #[cfg(not(target_arch = "wasm32"))]
+        tokio::task::spawn(async move {
+            gateway.gateway_listen_task().await;
+        });
+        #[cfg(target_arch = "wasm32")]
+        wasm_bindgen_futures::spawn_local(async move {
+            gateway.gateway_listen_task().await;
+        });
+
+        Ok(VoiceGatewayHandle {
+            url: websocket_url.clone(),
+            events: shared_events,
+            websocket_send: shared_websocket_send.clone(),
+            kill_send: kill_send.clone(),
+        })
+    }
+
+    /// The main gateway listener task;
+    ///
+    /// Can only be stopped by closing the websocket, cannot be made to listen for kill
```

Comment:
Not that this is really relevant IMO, but can it really not be made to listen for kill? My naive instinct would be to pass a kill message through a channel

Reply:
Perhaps not that it strictly _can't_ listen for the existing kill channel, it's just a lot easier to use the closing of the websocket

The task runs in a loop, so it's almost always awaiting a new websocket message, such that when the websocket closes it halts the loop.

We could use tokio::select to await messages on the kill channel or the websocket, but that would make the code unnecessarily more complex imo